Mathlib/CategoryTheory/Monoidal/Comon_.lean

```diff
@@ -0,0 +1,173 @@
+
+-/
+
+set_option linter.uppercaseLean3 false
+
+universe v₁ v₂ u₁ u₂ u
+
+open CategoryTheory MonoidalCategory
+
+variable (C : Type u₁) [Category.{v₁} C] [MonoidalCategory.{v₁} C]
+
+/-- A comonoid object internal to a monoidal category.
+
+When the monoidal category is preadditive, this is also sometimes called a "coalgebra object".
+-/
+structure Comon_ where
+  /-- The underlying object of a comonoid object. -/
+  X : C
+  /-- The counit of a comonoid object. -/
+  counit : X ⟶ 𝟙_ C
+  /-- The comultiplication morphism of a comonoid object. -/
+  comul : X ⟶ X ⊗ X
+  counit_comul : comul ≫ (counit ▷ X) = (λ_ X).inv := by aesop_cat
+  comul_counit : comul ≫ (X ◁ counit) = (ρ_ X).inv := by aesop_cat
+  comul_assoc : comul ≫ (comul ▷ X) = comul ≫ (X ◁ comul) ≫ (α_ X X X).inv := by aesop_cat
+
+attribute [reassoc] Comon_.counit_comul Comon_.comul_counit
+
+-- We prove a more general `@[simp]` lemma below.
+attribute [reassoc (attr := simp)] Comon_.comul_assoc
+
+namespace Comon_
+
+/-- The trivial comonoid object. We later show this is initial in `Comon_ C`.
+-/
+@[simps]
+def trivial : Comon_ C where
+  X := 𝟙_ C
+  counit := 𝟙 _
+  comul := (λ_ _).inv
+  comul_assoc := by coherence
+  counit_comul := by coherence
+  comul_counit := by coherence
+
+instance : Inhabited (Comon_ C) :=
+  ⟨trivial C⟩
+
+variable {C}
+variable {M : Comon_ C}
+
+@[simp]
+theorem counit_comul_hom {Z : C} (f : M.X ⟶ Z) : M.comul ≫ (M.counit ⊗ f) = f ≫ (λ_ Z).inv := by
+  rw [leftUnitor_inv_naturality, tensorHom_def, counit_comul_assoc]
+
+@[simp]
+theorem mul_one_hom {Z : C} (f : M.X ⟶ Z) : M.comul ≫ (f ⊗ M.counit) = f ≫ (ρ_ Z).inv := by
+  rw [rightUnitor_inv_naturality, tensorHom_def', comul_counit_assoc]
+
+theorem assoc_flip :
+    M.comul ≫ (𝟙 M.X ⊗ M.comul) = M.comul ≫ (M.comul ⊗ 𝟙 M.X) ≫ (α_ M.X M.X M.X).hom := by simp
+
+/-- A morphism of comonoid objects. -/
+@[ext]
+structure Hom (M N : Comon_ C) where
+  /-- The underlying morphism of a morphism of comonoid objects. -/
+  hom : M.X ⟶ N.X
+  hom_counit : hom ≫ N.counit = M.counit := by aesop_cat
+  hom_comul : hom ≫ N.comul = M.comul ≫ (hom ⊗ hom) := by aesop_cat
+
+attribute [reassoc (attr := simp)] Hom.hom_counit Hom.hom_comul
+
+/-- The identity morphism on a comonoid object. -/
+@[simps]
+def id (M : Comon_ C) : Hom M M where
+  hom := 𝟙 M.X
+
+instance homInhabited (M : Comon_ C) : Inhabited (Hom M M) :=
+  ⟨id M⟩
+
+/-- Composition of morphisms of monoid objects. -/
+@[simps]
+def comp {M N O : Comon_ C} (f : Hom M N) (g : Hom N O) : Hom M O where
+  hom := f.hom ≫ g.hom
+
+instance : Category (Comon_ C) where
+  Hom M N := Hom M N
+  id := id
+  comp f g := comp f g
+
+@[ext] lemma ext {X Y : Comon_ C} {f g : X ⟶ Y} (w : f.hom = g.hom) : f = g := Hom.ext _ _ w
+
+@[simp] theorem id_hom' (M : Comon_ C) : (𝟙 M : Hom M M).hom = 𝟙 M.X := rfl
+
+@[simp]
+theorem comp_hom' {M N K : Comon_ C} (f : M ⟶ N) (g : N ⟶ K) :
+    (f ≫ g : Hom M K).hom = f.hom ≫ g.hom :=
```

Comment:
```suggestion
    (f ≫ g).hom = f.hom ≫ g.hom :=
```
should be sufficient, unless you added the annotation for a specific reason?

Reply:
Yeah, doesn't seem necessary.